### ISSUE TITLE
Supplied key (org.apache.sshd.common.config.keys.OpenSshCertificateImpl) is not a RSAPublicKey instance

### ISSUE THREAD

**Issue by liuziyu1226** [apache/mina-sshd#650]
### Version

2.8.0

### Bug description

I overwrote the authenticate function of the PublickeyAuthenticator.java and registered it with the ssh server.

```
sshServer.setPublickeyAuthenticator(publicKeyAuth);
```
But when the client provides the key and certificate, MINA SSHD displays an error:

```
2024-12-25 17:29:30.882 sshd-SshServer[53600a30](port=2222)-nio2-thread-5 WARN  org.apache.sshd.server.session.ServerUserAuthService.warn - handleUserAuthRequestMessage(ServerSessionImpl[null@/127.0.0.1:56562]) Failed (InvalidKeyException) to authenticate using factory method=publickey: Supplied key (org.apache.sshd.common.config.keys.OpenSshCertificateImpl) is not a RSAPublicKey instance
```

After debugging, I found that the problem is that when MINA SSHD initializes the signature verifier of the public key, the public key passed in is of type OpenSshCertificateImpl, while Signature expects an RSAPublicKey object. Why does this problem occur? How can I solve it?

### Actual behavior

I rewrote the initVerifier function in SignatureRSA.java and now I can login correctly.
Old initVerifier function:
```
public void initVerifier(SessionContext session, PublicKey key) throws Exception {
      super.initVerifier(session, key);
      RSAKey rsaKey = ValidateUtils.checkInstanceOf(key, RSAKey.class, "Not an RSA key");
      verifierSignatureSize = getVerifierSignatureSize(rsaKey);
}
```
I overwrote the initVerifier function:
```
public void initVerifier(SessionContext session, PublicKey key) throws Exception {
      if (key instanceof OpenSshCertificate){
          super.initVerifier(session, ((OpenSshCertificate) key).getCertPubKey());
          RSAKey rsaKey = ValidateUtils.checkInstanceOf(((OpenSshCertificate) key).getCertPubKey(), RSAKey.class, "Not an RSA key");
          verifierSignatureSize = getVerifierSignatureSize(rsaKey);
      }else {
          super.initVerifier(session, key);
          RSAKey rsaKey = ValidateUtils.checkInstanceOf(key, RSAKey.class, "Not an RSA key");
          verifierSignatureSize = getVerifierSignatureSize(rsaKey);
      }
  }
```
I am still curious as to why this happens？ And whether there are any hidden dangers in my modification.
`

**Comment by tomaswolf** [comment]
The server side of user authentication with OpenSshCertificates was not fully implemented, and apparently is lacking tests. 

See issue [SSHD-1161](https://issues.apache.org/jira/browse/SSHD-1161).

The real cause of the problem you encounter is in

https://github.com/apache/mina-sshd/blob/7b19f6b678a72bb1ac02a8a852ec301d6791024c/sshd-core/src/main/java/org/apache/sshd/server/auth/pubkey/UserAuthPublicKey.java#L95

and

https://github.com/apache/mina-sshd/blob/7b19f6b678a72bb1ac02a8a852ec301d6791024c/sshd-core/src/main/java/org/apache/sshd/server/auth/pubkey/UserAuthPublicKey.java#L132

If the key is an OpenSshCertificate, `initVerifier` should be called with `key.getCertPubKey()`, not with `key`.

So, to answer your questions: it happens because there is a bug. There are no dangers to your modification, but it'll work only for RSA keys. The real fix has to be in `UserAuthPublicKey` (server side) so that it works for all key types. Similar to how it's done in https://github.com/apache/mina-sshd/blob/7b19f6b678a72bb1ac02a8a852ec301d6791024c/sshd-core/src/main/java/org/apache/sshd/client/kex/DHGClient.java#L184-L186 and https://github.com/apache/mina-sshd/blob/7b19f6b678a72bb1ac02a8a852ec301d6791024c/sshd-core/src/main/java/org/apache/sshd/client/kex/DHGClient.java#L222 Plus there needs to be a test for this.

Note that there are other bits of authenticating with a certificate missing on the server side: https://github.com/apache/mina-sshd/blob/7b19f6b678a72bb1ac02a8a852ec301d6791024c/sshd-core/src/main/java/org/apache/sshd/server/auth/pubkey/UserAuthPublicKey.java#L108-L110

**Comment by liuziyu1226** [comment]
Yes, the fundamental problem is that when initializing Signature, instead of getting cert publickey from OpenSshCertificate, OpenSshCertificate is directly passed in as the public key.
I tested the above method and found two problems:
1. It can only adapt to certs with RSA encryption algorithm
2. Even if the initVerifier() function of all encryption algorithms is rewritten, the signature will fail when the encryption method of cert is different from that of cert publickey.
Can I please raise a PR to fix the issue in UserAuthPublicKey?

**Comment by tomaswolf** [comment]
Sure, if you can provide a PR with the fix and tests with different certificates/key types, that'd be great.

**Comment by liuziyu1226** [comment]
OK, I will write my PR in the next few weeks after work, and I hope my work can make a small contribution to this open source project.

**Comment by tomaswolf** [comment]
No, from this info I cannot tell what might be wrong. Try debugging. The changes done for OpenSSH user certificate include unit tests that successfully authenticate at a server.

One thing: what do you upload as public key at the Gerrit server? There would need to be a way to upload the CA public key (not the certificate itself!) to Gerrit, and tell Gerrit that it is a CA pubkey. Gerrit's `PublickeyAuthenticator` would then have to check the certificates CA key against that uploaded CA key.

**Comment by calderonth** [comment]
Yes, I have of course modified gerrit to use the CA keys so that the validation happen and this seems to work as per logs:

```
    try {
      Path authorizedKeys = FileSystems.getDefault().getPath("/tmp/" , "ssh_user_key.pub");
      PublickeyAuthenticator authenticator = 
        new AuthorizedKeyEntriesPublickeyAuthenticator(
          authorizedKeys,
          session,
          AuthorizedKeyEntry.readAuthorizedKeys(authorizedKeys),
          PublicKeyEntryResolver.FAILING
        );
      return authenticator.authenticate(username, suppliedKey, session);
```
With the format being:
```
cert-authority ecdsa-sha2-nistp256 AAAAE2...
```

However, somehow the session seems to be in a broken state.
